### ISSUE TITLE
Is this python framework compatible with version 2 of the AHK binary yet?

### ISSUE THREAD

**Issue by gtusr** [spyoungtech/ahk#183]
[Expected release date of AHK v2](https://www.autohotkey.com/v2/)

A release candidate for v2 has been available since 11/20/22.

**Comment by spyoungtech** [comment]
No, AHK v2 is not yet supported. But support will be planned once AHKv2 is officially released.

**Comment by RePlay-be** [comment]
AHK v2 has been officially released. :)

https://www.autohotkey.com/v2/
https://www.autohotkey.com/docs/v2/ChangeLog.htm

**Comment by spyoungtech** [comment]
Coming back around to this, I'm wondering what is really involved to be 'compatible' with AHKv2. The only functionality that should be impacted by the version of AHK is when calling `run_script`, which is basically just a thin wrapper to call `subprocess.Popen` and pass the script. Otherwise, all interactions with AHK code are completely abstracted by the API.

Besides calling `run_script` with a v2 script, are there any other use-cases for AHKv2 compatibility?

**Comment by gtusr** [comment]
AHKv1.1 is now officially deprecated:

[Official AHK Download Page](https://www.autohotkey.com/)

That's a problem for many folks from a security point of view.

**Comment by spyoungtech** [comment]
Deprecated doesn't imply v1 doesn't receive security or bug fixes. Just no new features. The maintainers have committed to maintaining v1, just on a lower priority than v2, as described in the [announcement post](https://www.autohotkey.com/boards/viewtopic.php?f=24&t=112989&p=503343&hilit=deprecated#p503343).

So, from a security standpoint, there shouldn't be any issues while v1 is still maintained.

**Comment by gtusr** [comment]
I think the issue is that the main developers will not be spending much time maintaining v1 moving forward.  They will still accept patches from volunteers, but the core developers will be devoting most of their time to maintaining, enhancing and optimizing v2, which importantly includes keeping it up to date with changes to the underlying OS as new versions of Windows are released by Microsoft.

I've had a few AHK scripts break because of updates to the Windows API. Short term, you're probably right that AHKv1 compatibility is good enough.  Longer term, I think planning for a switch to v2 makes sense, especially since there are already scripts available on the internet that help automate updating v1 scripts to v2 syntax.

How difficult do you expect making a transition from v1 to v2 will be?

**Comment by spyoungtech** [comment]
I don't see there being a transition _from v1 to v2_ -- but rather, just adding support for v2.

This will probably entail the following:

1. Support the automatic discovery of multiple AHK binaries (e.g., as in [`_resolve_executable_path`](https://github.com/spyoungtech/ahk/blob/d07c45bf91769528316bc081310c6efff9e5def8/ahk/_async/transport.py#L288))
2. Detect the version of a provided AHK binary (and/or allow the user to specify the version of a binary provided by the `executable_path` keyword argument)
3. Detect the version of AHK required for user-provided scripts and use the appropriate binary automatically -- for example, when calling `run_script`. This may look something like what the official [Launcher](https://www.autohotkey.com/docs/v2/Program.htm#launcher) does (or maybe we leverage the launcher, somehow).
4. Perhaps most obviously, we will need to implement v2-compatible versions of the existing [autohotkey script templates](https://github.com/spyoungtech/ahk/tree/d07c45bf91769528316bc081310c6efff9e5def8/ahk/templates)

There may also be some other caveats to resolve, such as interface differences between v1 and v2 -- different parameters required or functionality that is exclusive to one version or the other and make sure the API facade handles this gracefully where possible.

**Comment by spyoungtech** [comment]
As it turns out, the vast majority of the existing code is actually compatible with both v1 and v2. Therefore, adding v2 support actually seems like it can be done in a short-term effort. v2 support will be available in an upcoming release 🎉

**Comment by spyoungtech** [comment]
Initial support for AutoHotkey V2 is being actively developed in #239 and will be released as a preview/beta feature in the next version.

**Comment by spyoungtech** [comment]
v2 support is currently available in pre-release version v1.4.0rc2

You can install it with `pip install ahk==1.4.0rc2`.

Use it like:

```python
from ahk import AHK

ahk = AHK(executable_path=r'C:\path\to\ahkv2.exe', version='v2')

ahkv2_code = '''\
#Requires AutoHotkey >= 2.0-
MsgBox("Hello v2")
'''
ahk.run_script(ahkv2_code)
```
If you install the latest version of `ahk-binary` you can omit the `executable_path` argument and the correct version of ahk should be used simply by using `version='v2'`.

**Comment by spyoungtech** [comment]
V2 support is now released.
